### PR TITLE
[gc] Speed up by improving SSTable parser and improving partitioning

### PR DESCRIPTION
**Pull request by arielshaqed**
1. Replace SparkSQL antijoin computations with a more Hadoopish Map/Reduce style that computes first differences of range IDs and then of object addresses.  This enables explicit control over partitioning.  Use hash-based partitioning (of both range IDs and object addresses!) to significantly reduce shuffling.
1. Speed up SSTable parser by replacing the use of ByteBuffers with plain byte arrays.  (This is _not_ a major speedup unfortunately, however by giving *Java* GC more visibility it _might_ improve memory usage.

This is _much_ faster (no precise timing yet, but a large repo got GCed in <2h with this code and never finished with the existing code, so this might be 10x?).

### CODE REVIEW

**Comment by arielshaqed** [comment]
# Hints for reviewers

Apologies for this long PR.  It's actually >= 2 separate PRs, that cannot be easily be separated.  This is my fault.

But you _can_ review it in smaller parts.

* GarbageCollector.scala.  Here we replaced the Spark-SQL distinct + antijoin operations with explicit Map/Reduce operations.  This allows precise control over partitioning, so we select hash-based partitioning and explicitly request the number of partitions.
* GarbageCollectorSpec.scala.  These are Spark component tests for GarbageCollector.  This verifies some behaviour of the Spark code by running it on an actual Spark (local) cluster.  **PLEASE consider whether I should add any tests here!**
* SSTableReader, BlockParser, IndexedBytes, and their tests.  Move from parsing on ByteBuffers (directly memory-mapped from the file and/or Snappy decompressor) to Array[Byte] (just plain Java byte arrays).  While ByteBuffers should be faster, in practice they aren't -- probably because we don't do enough with them.
* DumpSSTable.  A benchmark for the SSTable reader.
* run-gc-test.  Since GarbageCollector lets us select number of partitions, drop the number of partitions used.  This does not appear to speed up the test significantly 😿 .

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala
Comment (talSofer) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:131]: for my understanding, what is the reducer function? I can't understand by reading it. 
I expect it to generate 1 for items that are only part of left, but what would it generate for the rest?

File: clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSContext.scala
Comment (talSofer) [inline clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSContext.scala:24]: I suggest that we document it as a configuration option but we can do it under a different pr.

**Review by talSofer — COMMENTED** [review]
Thanks Ariel! 
I reviewed the GC part which looks great. and added some questions. posting this without approving because I have not reviewed the reader piece yet. 

A general question, do we know why things weren't partitioned properly in our previous code? or it is a mix of many things?

File: clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSContext.scala
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSContext.scala:24]: You're right, of course!  I'll start that PR (don't want to delay this one).

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala]: `_ | _` is Scala for "the bitwise OR function".  So it returns 1 for keys which have only been seen on the left, 2 for keys which have only been seen on the right, and 3 (== 1|2) for keys which have been seen on both sides.  On the next action we filter to keep only 1's -- keys which have only been seen on the left.  What can I say, this is a great time for [bit twiddling](http://catb.org/jargon/html/B/bit-twiddling.html).

Adding a name for `_ | _`, which I hope will make things understandable.

**Review by arielshaqed — COMMENTED** [review]
Thanks for the review!

The previous version didn't offer very wide parallelism anywhere: a single CSV file is a single partition, therefore mapping its commit ID to metarange IDs still yields a single partition, and splitting each into metarange into ranges is _still_ a single partition. Adding various "repartition" calls does increase parallelism, but does nothing for performance if you need to shuffle.

So seeing small number of partitions and a lot of shuffling when repartitioning without consideration to _what_ we're doing gives bad stats (low parallelism, a lot of shuffling if repartitioning).

The classic SparkSQL method of fixing this is by adding hints that say how to repartition each column.  By now we're at about the same code complexity of just coding a reduction.  If I want the critical part of the job to have a particular shape, I may as well write it out explicitly.

Note that GarbageCollector.scala is actually (much) _shorter_ in this version.  The PR adds code mainly because of tests.

File: clients/spark/core/src/test/scala/io/treeverse/clients/GarbageCollectorSpec.scala
Comment (johnnyaug) [inline clients/spark/core/src/test/scala/io/treeverse/clients/GarbageCollectorSpec.scala:34]: Why is this test relevant?

File: clients/spark/core/src/test/scala/io/treeverse/clients/GarbageCollectorSpec.scala
Comment (johnnyaug) [inline clients/spark/core/src/test/scala/io/treeverse/clients/GarbageCollectorSpec.scala:69]: Cool!

**Review by johnnyaug — APPROVED** [review]
Amazing work! 
Love the naming changes and the tests. Leaving some minor comments.

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:97]: Thanks!

It's a real annoyance: otherwise it's really hard to `import spark.implicits._`, and _nothing works_.

File: clients/spark/core/src/main/scala/io/treeverse/jpebble/BlockReadable.scala
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/jpebble/BlockReadable.scala]: Done!  I just removed it entirely, it's a pointless check anyway.

File: clients/spark/core/src/main/scala/io/treeverse/jpebble/IndexedBytes.scala
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/jpebble/IndexedBytes.scala:24]: ByteBufferIndexedBytesSpec tests this as part of "iterator".

File: clients/spark/core/src/test/scala/io/treeverse/clients/GarbageCollectorSpec.scala
Comment (arielshaqed) [reply clients/spark/core/src/test/scala/io/treeverse/clients/GarbageCollectorSpec.scala:34]: It demonstrates how to test using Spark.  Otherwise it's really hard to understand what an easy form is.  Opened #4211 to do it soon-ish.

**Review by arielshaqed — COMMENTED** [review]
Thanks!

Waiting for review by @Jonathan-Rosenberg before pulling.

**Review by Jonathan-Rosenberg — APPROVED** [review]
Went through the parser part. LGTM.

**Comment by arielshaqed** [comment]
Pulling once tests pass (silly failure).

**Pull request merged**
